{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fceea6cb",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.Seq import Seq\n",
    "from string import ascii_letters\n",
    "from Bio import SeqIO, SeqUtils\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d26154bc",
   "metadata": {},
   "outputs": [],
   "source": [
    "class BiStudent:\n",
    "    rus_letters = 'абвгдеёжзийклмнопрстуфхцчшщьыъэюя-'\n",
    "    rus_letters_up = rus_letters.upper()\n",
    "\n",
    "    def __init__(cls, fullname, age, sex, program):\n",
    "        \"\"\"There are following arguments:\n",
    "        -fullname: str with 2 or more words in latin or cyrillic\n",
    "\n",
    "        -age: between 18 and 40\n",
    "\n",
    "        -sex: one of the 'Male', 'M', 'Female', 'F'\n",
    "\n",
    "        -program: one of the 'Bioinformatics for Bioilogist', 'Bio', 'Algorithmic Bioinformatics', 'Inf' \"\"\"\n",
    "        cls.verify_fullname(fullname)\n",
    "        cls.verify_age(age)\n",
    "        cls.verify_sex(sex)\n",
    "        cls.verify_program(program)\n",
    "\n",
    "        cls.__fullname = fullname.strip()\n",
    "        cls.age = age\n",
    "        cls.sex = sex\n",
    "        cls.program = program\n",
    "\n",
    "    def __str__(cls):\n",
    "        return \"The class BiStudent represents some information about a student\"\n",
    "\n",
    "    def get_info(cls):\n",
    "        return f\"{cls.__fullname} is {cls.age} y.o and she/he is a {cls.sex}. \" \\\n",
    "               f\"She/he studies in {cls.program} program\"\n",
    "\n",
    "    @classmethod\n",
    "    def verify_fullname(cls, fullname):\n",
    "        if not isinstance(fullname, str):\n",
    "            raise TypeError('Fullname must be a str type')\n",
    "        fio = fullname.split()\n",
    "        if len(fio) != 3 and len(fio) != 2:\n",
    "            raise TypeError('Fullname must contain: name middlename (optional) and lastname')\n",
    "        letters = cls.rus_letters + cls.rus_letters_up + ascii_letters\n",
    "        for s in fio:\n",
    "            if len(s) < 1:\n",
    "                raise TypeError('Length of any word in fullname must be higher than 1')\n",
    "            if len(s.strip(letters)) != 0:\n",
    "                raise TypeError('Invalid symbols are received')\n",
    "\n",
    "    @classmethod\n",
    "    def verify_age(cls, age):\n",
    "        if not isinstance(age, int) or age < 18 or age > 40:\n",
    "            raise TypeError('Age must be between 18 and 40 an int type')\n",
    "\n",
    "    @classmethod\n",
    "    def verify_sex(cls, sex):\n",
    "        if sex not in ('Male', 'M', 'Female', 'F'):\n",
    "            raise TypeError('the correct sex: (Male, M, Female, F)')\n",
    "\n",
    "    @classmethod\n",
    "    def verify_program(cls, program):\n",
    "        if program not in ('Bioinformatics for Bioilogist', 'Bio', 'Algorithmic Bioinformatics', 'Inf'):\n",
    "            raise TypeError(\"the correct program: ('Bioinformatics for Bioilogist', 'Bio',\"\n",
    "                            \" 'Algorithmic Bioinformatics', 'Inf')\")\n",
    "\n",
    "    @property\n",
    "    def fullname(cls):\n",
    "        return cls.__fullname\n",
    "\n",
    "    @staticmethod\n",
    "    def love_dogs(flag=True):\n",
    "        \"\"\"It draws a plot depends on student's preferences\"\"\"\n",
    "        if flag:\n",
    "            sns.dogplot()\n",
    "            plt.show()\n",
    "        else:\n",
    "            print('If you liked dogs you would get a special prize. Unfortunately, you don\\'t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "682d7832",
   "metadata": {},
   "outputs": [],
   "source": [
    "class Rna:\n",
    "    def __init__(cls, seq):\n",
    "        \"\"\"This class works with Rna sequences. Takes one argumens: 'seq' - Rna sequence\"\"\"\n",
    "        if not set(seq.upper()).issubset({'A', 'U', 'G', 'C', 'N'}):\n",
    "            raise TypeError(\"RNA sequence must contain 'A' 'U' 'G' 'C' 'N' case insensitive\")\n",
    "        cls.seq = seq\n",
    "\n",
    "    def __str__(cls):\n",
    "        return \"Class Rna works with rna-molecules. It can return translated protein or complement cDNA\"\n",
    "\n",
    "    def translate_rna(cls):\n",
    "        \"\"\"This function performs rna->protein translation\"\"\"\n",
    "        translate = Seq(cls.seq).translate()\n",
    "        if '*' in translate:\n",
    "            print('* - means stop-codon')\n",
    "        return f\"Protein-sequence: {translate}\"\n",
    "\n",
    "    def back_transcribe(cls):\n",
    "        \"\"\"This function returns reverse complement cDNA\"\"\"\n",
    "        return Seq(cls.seq).back_transcribe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "73efdc66",
   "metadata": {},
   "outputs": [],
   "source": [
    "class PositiveSet(set):\n",
    "    \"\"\"This class changes set built-in data-structure. Takes one argument - 'collection'\"\"\"\n",
    "\n",
    "    def __init__(cls, collection=()):\n",
    "        collection = filter(lambda x: x >= 0, collection)\n",
    "        super().__init__(collection)\n",
    "\n",
    "    def add(cls, num):\n",
    "        \"\"\"This function adds positive number to a PositiveSet object\"\"\"\n",
    "        if num >= 0:\n",
    "            super().add(num)\n",
    "        else:\n",
    "            raise ValueError('It only adds positive numbers')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d62396d3",
   "metadata": {},
   "outputs": [],
   "source": [
    "class FastaStat:\n",
    "    def __init__(cls, path_to_fasta):\n",
    "        \"\"\"Only one required argument - path_to fasta. The absolute or relative path to a .fasta file\"\"\"\n",
    "        cls._fasta = cls._read_fasta(path_to_fasta)\n",
    "        cls.path = path_to_fasta\n",
    "\n",
    "    def __str__(cls):\n",
    "        return os.path.abspath(cls.path)\n",
    "\n",
    "    @classmethod\n",
    "    def _read_fasta(cls, path):\n",
    "        \"\"\"Reads fasta file and save it to fasta variable\"\"\"\n",
    "        fasta = list(SeqIO.parse(path, 'fasta'))\n",
    "        return fasta\n",
    "\n",
    "    def n_reads(cls):\n",
    "        \"\"\"Return number of reads in fasta file accordingly to SeqIO.\"\"\"\n",
    "        return len(cls._fasta)\n",
    "\n",
    "    def length_distribution(cls):\n",
    "        \"\"\"Make a barplot of length distribution over the reads\"\"\"\n",
    "        lengths = {}\n",
    "        counter = 0\n",
    "        for record in cls._fasta:\n",
    "            lengths[counter] = len(record.seq)\n",
    "            counter += 1\n",
    "        plt.bar(x=lengths.keys(), height=lengths.values(), color='blue', alpha=0.4, edgecolor='black', linewidth=3)\n",
    "        plt.rcParams[\"figure.figsize\"] = (12, 8)\n",
    "        plt.xlabel('Read number', size=16)\n",
    "        plt.ylabel('Read length', size=16)\n",
    "        plt.title('Length distribution over the reads', size=20)\n",
    "        plt.show()\n",
    "\n",
    "    def gc_content(cls, for_all=False):\n",
    "        \"\"\"Return GC-content average over all reads.\n",
    "        If for_all=True returns GC-content in each read.\n",
    "        \"\"\"\n",
    "        if for_all:\n",
    "            for record in cls._fasta:\n",
    "                print(f'{record.id}:   {round(SeqUtils.GC(record.seq), 2)}%')\n",
    "            return 'Done'\n",
    "        gc = 0\n",
    "        for record in cls._fasta:\n",
    "            gc += SeqUtils.GC(record.seq)\n",
    "        return f'GC-content: {round(gc / len(cls._fasta), 2)}%'\n",
    "\n",
    "    def four_mers_count(cls):\n",
    "        \"\"\"Return 4-mers distribution plot.\"\"\"\n",
    "        seq_four_mers = {}\n",
    "        for record in cls._fasta:\n",
    "            current_seq = record.seq\n",
    "            for n in range(len(current_seq[:-3])):\n",
    "                four_mer = str(current_seq[n:n + 4])\n",
    "                if four_mer not in seq_four_mers:\n",
    "                    seq_four_mers[four_mer] = 1\n",
    "                else:\n",
    "                    seq_four_mers[four_mer] += 1\n",
    "        plt.bar(x=seq_four_mers.keys(), height=seq_four_mers.values(), color='blue', alpha=0.4, edgecolor='black',\n",
    "                linewidth=1)\n",
    "        plt.rcParams[\"figure.figsize\"] = (20, 8)\n",
    "        plt.rcParams['figure.dpi'] = 150\n",
    "        plt.xlabel('4 mer', size=16)\n",
    "        plt.ylabel('4 mer count', size=16)\n",
    "        plt.title('4 mers distribution', size=20)\n",
    "        plt.xticks(rotation=90, size=6)\n",
    "        plt.show()\n",
    "\n",
    "    def execute_everything(cls):\n",
    "        filtered_attr = list(filter(lambda x: not x.startswith('_'), dir(cls)))\n",
    "        callable_attr = [method for method in filtered_attr if callable(getattr(cls, method))]\n",
    "        callable_attr.remove('execute_everyhing')\n",
    "        for attr in callable_attr:\n",
    "            obj = getattr(cls, attr)\n",
    "            print(obj())"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
